### PR TITLE
Milestone 3

### PR DESCRIPTION
**Pull request by abouhid**
This PR is being created once more because the submission is not accepting the previous ones. The issue here is because we submitted milestones 2, 3, and 4 at the same time (hence the milestone 4 project is already approved in our dashboard). For some reason the third milestone was not accepted as follows.
image

### CODE REVIEW

**Comment by sumancrest0001** [comment]
**Status: Changes Required :gear: 
Hello @abouhid  and @KelynPNjeri,** It is a little bit confusing as there is no comment from the previous TSE. Did you create two PR for this milestone? You haven't implemented any requirements from this milestone. Let's consider following points and add features accordingly. In the end, friendship features should be very similar to Facebook.
1. Current, I can only add friend to another user but there are not options for confirmation, cancelling requests. [This](https://explainextended.com/2009/03/07/selecting-friends/) article can be useful.
2. Now, you should render post from the current_user and friends only. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/timeline.png)
3. Also, add tests for the added features. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png)

**Review by sumancrest0001 — CHANGES_REQUESTED** [review]
**Please ensure that above features are added in your project and submit for another review.
If you have queries regarding the review, message me on slack Suman Shrestha.
Happy coding !!!**

**Comment by sumancrest0001** [comment]
**I am extremely sorry for the mistake. Please submit for another review.**

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      redirect_to users_path, notice: @friendship_created.errors.full_messages
+    end
+  end
+
+  def update
+    @friendship = Friendship.find(params[:id])
+    @friendship.confirmed = true
+    @inverse_friendship = Friendship.create(friend_id: @friendship.user_id,
+                                            user_id: @friendship.friend_id,
+                                            confirmed: true)
```
Comment (SpaYco) [inline app/controllers/friendships_controller.rb]: - [ ] move this to the friendship model as a method

```Ruby
def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

* [ ] There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers, view-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb), there should be absolutely NO logic in views.

* [ ]  we can find user friends only by user_id in friendships table

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

* [ ] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

* [ ] User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

* [ ] in order to get all posts written by user on their friends, you can use simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

_P.S: you can use [this artcile](https://explainextended.com/2009/03/07/selecting-friends) and [stackoverflow discussion](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) as hints._


### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello Team
### Good job so far :clap: However, the branch has some conflicts that you might need to solve before we proceed. 
### Make the changes and submit for another code review. Good luck.
- Happy coding!

**Review by rOluochKe — APPROVED** [review]
Hello Team
### You got it right, well done :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

**Pull request merged**